### PR TITLE
Master keyboard navigation vsc

### PR DESCRIPTION
**Pull request by VincentSchippefilt**
Description of the issue/feature this PR addresses:
https://www.odoo.com/web#id=29653&view_type=form&model=project.task&action=333&active_id=133&menu_id=4720

Current behavior before PR:
When using tab navigation, the user could not enter many2X fields.

Desired behavior after PR is merged:
It is possible to enter many2X fields using TAB key. Moving from field to field using tab will work as expected and will stop on required fields if they are not filled in.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web/static/src/js/chrome/abstract_web_client.js [hunk 1]
```diff
@@ -158,6 +167,54 @@ var AbstractWebClient = Widget.extend(ServiceProviderMixin, {
...
+                                usedAccessKey.push(candidateAccessKey);
+                                break;
+                            }
+                        }
+                    }
+                });
+                
+                var accesskeyElements = $(document).find('[accesskey]').filter(':visible');
+                _.each(accesskeyElements, function (elem) {
+                    var overlay = $(_.str.sprintf("<div class='o_web_accesskey_overlay'>%s</div>", $(elem).attr('accesskey').toUpperCase()));
```
Comment (ged-odoo) [inline addons/web/static/src/js/chrome/abstract_web_client.js]: $('<div>').addClass('o_web_...').text(...)

File: addons/web/static/src/js/chrome/abstract_web_client.js [hunk 2]
```diff
@@ -158,6 +167,54 @@ var AbstractWebClient = Widget.extend(ServiceProviderMixin, {
...
+                        overlay.appendTo($(elem).parent().css('position', 'relative'));
+                    }
+                    else {
+                        overlay.appendTo($(elem).css('position', 'relative'));
+                    }
+                });
+            } 
+        });
+        this.$el.on('keyup',function(event){
+            self.hide_accesskey_overlay.call(self);
```
Comment (ged-odoo) [inline addons/web/static/src/js/chrome/abstract_web_client.js]: self.hide_accesskey_overlay();

File: addons/web/static/src/js/views/abstract_controller.js [hunk 3]
```diff
@@ -102,7 +102,12 @@ var AbstractController = Widget.extend(ControlPanelMixin, {
         }
         return this._super.apply(this, arguments);
     },
-
+    /**
+     * Gives the focus to the renderer
+     */
+    giveFocus:function(){
+        this.renderer.giveFocus();
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/abstract_controller.js]: boom (discuss)

File: addons/web/static/src/js/fields/abstract_field.js [hunk 4]
```diff
@@ -299,6 +299,32 @@ var AbstractField = Widget.extend({
         return this._render() || $.when();
     },
 
+    /**
+     * add the invalid class on a field
+     */
+    setInvalidClass: function () {
```
Comment (ged-odoo) [inline addons/web/static/src/js/fields/abstract_field.js]: I know you don't like alphabetical order, and I am unsure about it, but maybe at least order these 3 methods

File: addons/web/static/src/js/fields/abstract_field.js [hunk 5]
```diff
@@ -299,6 +299,32 @@ var AbstractField = Widget.extend({
...
+        this.$el.toggleClass('o_field_invalid',true);
```
Comment (ged-odoo) [inline addons/web/static/src/js/fields/abstract_field.js]: maybe simpler to use this.$el.addClass('o_field_invalid')

File: addons/web/static/src/js/fields/abstract_field.js [hunk 6]
```diff
@@ -299,6 +299,32 @@ var AbstractField = Widget.extend({
...
+    },
+
+    /**
+     * Remove the invalid class on a field
+     */
+    removeInvalidClass: function () {
+        this.$el.toggleClass('o_field_invalid',false);
```
Comment (ged-odoo) [inline addons/web/static/src/js/fields/abstract_field.js]: and this.$el.removeClass('o_fieldInvalid')

File: addons/web/static/src/js/fields/abstract_field.js (hunk 4 above)
Comment (ged-odoo) [inline addons/web/static/src/js/fields/abstract_field.js]: oh, and this method looks a little bit too much 'implementation dependant'.  Maybe it should be renamed to something setInvalidState or something

File: addons/web/static/src/js/fields/abstract_field.js [hunk 7]
```diff
@@ -299,6 +299,32 @@ var AbstractField = Widget.extend({
...
+    },
+
+    /**
+     * Update the modifiers with the newest value. 
+     * Now this.attrs.modifiersValue can be used consistantly even with conditional modifiers
```
Comment (ged-odoo) [inline addons/web/static/src/js/fields/abstract_field.js]: would be useful to have a comment on why this is needed, or when this could be called.

File: addons/web/static/src/js/fields/abstract_field.js [hunk 8]
```diff
@@ -299,6 +299,32 @@ var AbstractField = Widget.extend({
...
+     * 
+     * @param {Object} modifiers  the updated modifiers
+     * @override
+     */
+    updateModifiersValue: function(modifiers) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/fields/abstract_field.js]: if you make the field widget aware of its modifier value, maybe the isValid method should use it and the rest of the system could be changed as well?

File: addons/web/static/src/js/fields/relational_fields.js [hunk 9]
```diff
@@ -794,6 +796,30 @@ var FieldX2Many = AbstractField.extend({
         return this._super.apply(this, arguments);
     },
 
+    /**
+     * @override
+     * @returns {jQuery}
+     */
+    getFocusableElement: function () {
+       return this.mode === 'edit' && this.$input || this.$el;
```
Comment (ged-odoo) [inline addons/web/static/src/js/fields/relational_fields.js]: I personally prefer parenthesis in this case: is a && b || c the same as (a && b) || c or a && (b || c) ? I don't know...

File: addons/web/static/src/js/fields/relational_fields.js [hunk 10]
```diff
@@ -1156,6 +1182,16 @@ var FieldX2Many = AbstractField.extend({
     _onToggleColumnOrder: function (ev) {
         ev.data.field = this.name;
     },
+    /*
+    * Move to next widget.
+    *
+    * @private
+    */
+    _onActiveNextWidget: function (e) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/fields/relational_fields.js]: missing doc string

File: addons/web/static/src/js/views/form/form_controller.js [hunk 11]
```diff
@@ -218,7 +235,32 @@ var FormController = BasicController.extend({
     //--------------------------------------------------------------------------
     // Private
     //--------------------------------------------------------------------------
-
+    /**
+     * Assign on the buttons save and discard additionnal behavior to facilitate the work of the users doing input only using the keyboard
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/form/form_controller.js]: 80 characters... it may be annoying to you, but i really prefer being able to read the doc

File: addons/web/static/src/js/views/form/form_controller.js [hunk 12]
```diff
@@ -218,7 +235,32 @@ var FormController = BasicController.extend({
...
+     * 
+     * @param {jQueryElement} $saveCancelButtonContainer  The div containing the save and cancel buttons
+     */
+    _assignSaveCancelKeyboardBehavior:function($saveCancelButtonContainer) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/form/form_controller.js]: should be tagged private.  also, not really important, but everywhere else, we add a space after the : and after function

File: addons/web/static/src/js/views/form/form_controller.js [hunk 13]
```diff
@@ -488,6 +530,16 @@ var FormController = BasicController.extend({
         event.stopPropagation();
         this.model.freezeOrder(event.data.id);
     },
+    _onFocusControlButton:function(e) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/form/form_controller.js]: missing docstring (in particular the type of e), should be tagged private

File: addons/web/static/src/js/views/form/form_controller.js [hunk 14]
```diff
@@ -488,6 +530,16 @@ var FormController = BasicController.extend({
...
+        if (this.$buttons) {
+            e.stopPropagation();
+            this.$buttons.find('.btn-primary:visible:first()').focus();
+        }
+    },
+    _onFormDialogDiscarded: function(e) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/form/form_controller.js]: same

File: addons/web/static/src/js/views/form/form_renderer.js [hunk 15]
```diff
@@ -43,10 +48,17 @@ var FormRenderer = BasicRenderer.extend({
...
      * first focusable field otherwise.
+     * In read mode, delegate which button to give the focus to, to the form_renderer
      */
     autofocus: function () {
         if (this.mode === 'readonly') {
-            return;
+            var firstPrimaryFormButton =  this.$el.find('button.oe_highlight:enabled:visible:first()');
+            if (firstPrimaryFormButton.length > 0) {
+                return firstPrimaryFormButton.focus();
+            }
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/form/form_renderer.js]: not much, but } else { should be on same line

File: addons/web/static/src/js/views/form/form_renderer.js [hunk 16]
```diff
@@ -43,10 +48,17 @@ var FormRenderer = BasicRenderer.extend({
...
      * Focuses the field having attribute 'default_focus' set, if any, or the
      * first focusable field otherwise.
+     * In read mode, delegate which button to give the focus to, to the form_renderer
      */
     autofocus: function () {
         if (this.mode === 'readonly') {
-            return;
+            var firstPrimaryFormButton =  this.$el.find('button.oe_highlight:enabled:visible:first()');
+            if (firstPrimaryFormButton.length > 0) {
+                return firstPrimaryFormButton.focus();
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/form/form_renderer.js]: if this function returns something, this should be documented

File: addons/web/static/src/js/views/form/form_renderer.js [hunk 17]
```diff
@@ -144,6 +156,14 @@ var FormRenderer = BasicRenderer.extend({
         this.$('.o_statusbar_buttons button, .oe_button_box button')
             .removeAttr('disabled');
     },
+    /**
+     * Put the focus on the last activated widget. 
+     * This function is used when closing a dialog to give the focus back to te form that has opened it and
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/form/form_renderer.js]: 80 chars

File: addons/web/static/src/js/views/kanban/kanban_controller.js [hunk 18]
```diff
@@ -220,6 +221,12 @@ var KanbanController = BasicController.extend({
             });
         }
     },
+    _onButtonsKeyDown: function(e) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/kanban/kanban_controller.js]: missing docstring, function should be tagged private,

File: addons/web/static/src/js/views/kanban/kanban_controller.js [hunk 19]
```diff
@@ -220,6 +221,12 @@ var KanbanController = BasicController.extend({
...
+        switch(e.keyCode) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/kanban/kanban_controller.js]: if would personally do a if instead of a switch

Comment (ged-odoo) [inline addons/web/static/src/js/views/kanban/kanban_controller.js]: if would personally do a if instead of a switch

File: addons/web/static/src/js/views/kanban/kanban_record.js [hunk 20]
```diff
@@ -504,6 +505,14 @@ var KanbanRecord = Widget.extend({
                 this.do_warn("Kanban: no action for type : " + type);
         }
     },
+    _onKanbanKeyDown: function (event) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/kanban/kanban_record.js]: docstring (in particular event)

File: addons/web/static/src/js/views/kanban/kanban_record.js [hunk 21]
```diff
@@ -504,6 +505,14 @@ var KanbanRecord = Widget.extend({
...
+        switch (event.keyCode) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/kanban/kanban_record.js]: i would do a if personally

File: addons/web/static/src/js/views/kanban/kanban_renderer.js [hunk 22]
```diff
@@ -88,6 +88,61 @@ var KanbanRenderer = BasicRenderer.extend({
         start_quick_create: '_onStartQuickCreate',
         quick_create_column_updated: '_onQuickCreateColumnUpdated',
     }),
+    events:_.extend({}, BasicRenderer.prototype.events || {}, {
+        'keydown .o_kanban_record' : '_onRecordKeyDown'
+    }),
+    _focusOnNextCard: function(currentCardElement) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/kanban/kanban_renderer.js]: private function, should be documented, should be in private section

File: addons/web/static/src/js/views/kanban/kanban_renderer.js [hunk 23]
```diff
@@ -88,6 +88,61 @@ var KanbanRenderer = BasicRenderer.extend({
...
+        var nextCard = currentCardElement.nextElementSibling;
+        if (nextCard) {
+            nextCard.focus();
+        }
+    },
+    _focusOnPrevousCard: function(currentCardElement) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/kanban/kanban_renderer.js]: private function, should be documented, should be in private section. Also, Previous, not Prevous

File: addons/web/static/src/js/views/kanban/kanban_renderer.js [hunk 24]
```diff
@@ -88,6 +88,61 @@ var KanbanRenderer = BasicRenderer.extend({
...
+        var previousCard = currentCardElement.previousElementSibling;
+        if (previousCard && previousCard.classList.contains("o_kanban_record")) { //previous element might be column title
+            previousCard.focus();
+        }
+    },
+    _onRecordKeyDown: function(e) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/kanban/kanban_renderer.js]: handler, should be documented, in handlers section

File: addons/web/static/src/js/views/list/list_controller.js [hunk 25]
```diff
@@ -221,6 +230,33 @@ var ListController = BasicController.extend({
             .toggleActive(ids, !archive, this.handle)
             .then(this.update.bind(this, {}, {reload: false}));
     },
+    /**
+     * Assign on the buttons create additionnal behavior to facilitate the work of the users doing input only using the keyboard
+     * 
+     * @param {jQueryElement} $createButton  The create button itself
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/list/list_controller.js]: should be tagged as private, 80 chars limit

File: addons/web/static/src/js/views/list/list_controller.js [hunk 26]
```diff
@@ -353,7 +389,9 @@ var ListController = BasicController.extend({
         // we prevent the event propagation because we don't want this event to
         // trigger a click on the main bus, which would be then caught by the
         // list editable renderer and would unselect the newly created row
-        event.stopPropagation();
+        if (event) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/list/list_controller.js]: why is it needed? if the method is sometimes called without an event, this code is not a handler and should be moved in a private function, and the handler will call it

File: addons/web/static/src/js/views/list/list_editable_renderer.js [hunk 27]
```diff
@@ -738,6 +739,34 @@ ListRenderer.include({
     _onFooterClick: function () {
         this.unselectRow();
     },
+    _onKeyDownAddRecord: function(e) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/list/list_editable_renderer.js]: docstring, tagged private, missing space after function

File: addons/web/static/src/js/views/list/list_editable_renderer.js [hunk 28]
```diff
@@ -738,6 +739,34 @@ ListRenderer.include({
...
+                e.preventDefault();
+                this._onAddRecord(e);
+                break;
+        }
+    },
+    /** 
+     * It will returns the first visible widget that is editable
+     *
+     * @private
+     * @returns {Class} Widget returns first widget
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/list/list_editable_renderer.js]: you can use the type Widget instead of class

Comment (ged-odoo) [reply addons/web/static/src/js/views/list/list_editable_renderer.js]: also, are we guaranteed to always have a return value? usually when we find something, null/undefined/-1 may happend...

File: addons/web/static/src/js/views/list/list_editable_renderer.js [hunk 29]
```diff
@@ -738,6 +739,34 @@ ListRenderer.include({
...
+        switch(e.keyCode) {
+            case $.ui.keyCode.ENTER:
+                e.stopPropagation();
+                e.preventDefault();
+                this._onAddRecord(e);
+                break;
+        }
+    },
+    /** 
+     * It will returns the first visible widget that is editable
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/list/list_editable_renderer.js]: It will return (no s)

File: addons/web/static/src/js/core/dialog.js [hunk 30]
```diff
@@ -184,11 +216,31 @@ var Dialog = Widget.extend({
...
+                // Keep class modal-open (deleted by bootstrap hide fnct) on body to allow scrolling inside the modal
+                $('body').addClass('modal-open');
+            }
+        }
+    },
+    /**
+     * Manages the focus when the dialog closes. The default behavior is to set the focus on the top-most opened popup.
+     * The goal of this function is to be overridden by all children of the dialog class.
+     * 
+     * @returns: bool  should return true if the focus has already been set else false.
```
[the remark this answers is not in the document]
Comment (ged-odoo) [reply addons/web/static/src/js/core/dialog.js]: also, bool is not a valid type, and this is not docstring compliant: {boolean}.  (this will probably break the doc build)

File: addons/web/static/src/js/widgets/auto_complete.js [hunk 31]
```diff
@@ -216,5 +216,8 @@ return Widget.extend({
     is_expandable: function () {
         return !!this.$('.o-selection-focus .o-expand').length;
     },
+    is_expanded: function() {
```
Comment (ged-odoo) [inline addons/web/static/src/js/widgets/auto_complete.js]: could you rename this to isExpanded?

Comment (ged-odoo) [reply addons/web/static/src/js/widgets/auto_complete.js]: and document it (it returns a boolean)

File: addons/web/static/src/less/kanban_dashboard.less [hunk 32]
```diff
@@ -43,6 +43,14 @@
             margin-bottom: 10px;
         }
 
+        &:focus, &:focus-within {
+            -webkit-box-shadow: 1px 1px 3px 0px rgba(0,0,0,1);
```
Comment (ged-odoo) [inline addons/web/static/src/less/kanban_dashboard.less]: use mixin: .box-shadow(...)

File: addons/web/static/src/less/kanban_view.less [hunk 33]
```diff
@@ -154,6 +154,13 @@
         @media (max-width: @screen-xs-max) {
             padding: @o-kanban-inside-vgutter @o-kanban-inside-hgutter-mobile;
         }
+        &:focus, &:focus-within {
+            -webkit-box-shadow: 3px 3px 10px 0px rgba(0,0,0,1);
```
Comment (ged-odoo) [inline addons/web/static/src/less/kanban_view.less]: use mixin

File: addons/web/static/src/less/list_view.less [hunk 34]
```diff
@@ -117,6 +117,18 @@
         border-top: none;  // Override bootstrap for grouped list views
     }
 
+    tbody {
+        tr:focus-within {
+            background-color: @odoo-form-lightsecondary;
+        }
+        tr{
+            .focussed {
```
Comment (ged-odoo) [inline addons/web/static/src/less/list_view.less]: http://www.future-perfect.co.uk/grammar-tip/is-it-focussed-or-focused/ (focused is highly preferred)

File: addons/web/static/src/less/webclient.less [hunk 35]
```diff
@@ -128,3 +128,18 @@ div.o_boolean_toggle {
...
+    height: 100%;
+    left: 0;
+    top: 0;
+    z-index: 1000000;
+    background-color: rgba(0,0,0,.6);
+    color: #FFFFFF;
+    justify-content: center;
+    display: flex;
+    align-items: center;
+}
```
Comment (ged-odoo) [inline addons/web/static/src/less/webclient.less]: missing empty line at the end (no idea why, but github displays this in red, so it should be fixed!

File: addons/web/static/tests/fields/relational_fields_tests.js [hunk 36]
```diff
@@ -11185,6 +11185,219 @@ QUnit.module('relational_fields', {
...
+
+        var form = createView({
+            View: FormView,
+            model: 'partner',
+            viewOptions: {
+                mode: 'edit',
+            },
+            data: this.data,
+            arch:'<form string="Partners">' +
+                    '<sheet>' +
```
Comment (ged-odoo) [inline addons/web/static/tests/fields/relational_fields_tests.js]: you can remove the sheet/notebook/pages tag, they should not influence the test

File: addons/web/static/tests/views/form_tests.js [hunk 37]
```diff
@@ -6717,6 +6723,366 @@ QUnit.module('Views', {
...
+                    '<field name="foo" />'+
+                '</form>',
+        });
+
+        form.$buttons.find('.o_form_buttons_edit').tooltip('show',false);
+        assert.strictEqual($('.tooltip .oe_tooltip_string').length, 1,
+            "should have rendered a tooltip");
+        form.destroy();
+    });
+    QUnit.test('if the focus is on the save button, hitting ENTER should save', function (assert) {
```
Comment (ged-odoo) [inline addons/web/static/tests/views/form_tests.js]: missing empty line to separate tests

File: addons/web/static/tests/views/form_tests.js [hunk 38]
```diff
@@ -6717,6 +6723,366 @@ QUnit.module('Views', {
...
+                }
+                return this._super(route, args);
+            },
+        });
+
+        form.$buttons.find('.o_form_button_save')
+                     .focus()
+                     .trigger($.Event('keydown', {which: $.ui.keyCode.ENTER}));
+        form.destroy();
+    });
```
Comment (ged-odoo) [inline addons/web/static/tests/views/form_tests.js]: need empty line to separate tests

File: addons/web_editor/static/src/js/editor/summernote.js [hunk 39]
```diff
@@ -1060,10 +1060,12 @@ $.summernote.pluginEvents.tab = function (event, editor, layoutInfo, outdent) {
             } else {
                 r = dom.merge(r.sc.parentNode, r.sc, r.so, r.ec, r.eo, null, true);
                 r = range.create(r.sc, r.so, r.ec, r.eo);
-                next = r.sc.splitText(r.so);
-                r.sc.textContent = r.sc.textContent.replace(/(\u00A0)+$/g, '');
-                next.textContent = next.textContent.replace(/^(\u00A0)+/g, '');
-                range.create(r.sc, r.sc.textContent.length, r.sc, r.sc.textContent.length).select();
+                if (r.sc.splitText) {
```
Comment (ged-odoo) [inline addons/web_editor/static/src/js/editor/summernote.js]: isn't this merged in master, or done by the web editor team?

File: odoo/tests/common.py [hunk 40]
```diff
@@ -445,7 +445,7 @@ def phantom_run(self, cmd, timeout):
             _logger.info('phantomjs unlink localstorage %s', i)
             os.unlink(i)
         try:
-            phantom = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=None, close_fds=True)
+            phantom = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=None)
```
Comment (ged-odoo) [inline odoo/tests/common.py]: no idea why your branch should change this. if this is necessary, should be in another commit with a good description of the issue in the commit message

File: addons/web/static/src/js/core/browser_detection.js [hunk 41]
```diff
@@ -0,0 +1,20 @@
+odoo.define('web.BrowserDetection', function (require) {
```
[the remark this answers is not in the document]
Comment (VincentSchippefilt) [reply addons/web/static/src/js/core/browser_detection.js]: If I leave it as a class, it can be extended or included. Not sure if it is necessary though.

File: addons/web/static/src/js/chrome/abstract_web_client.js [hunk 42]
```diff
@@ -100,6 +101,7 @@ var AbstractWebClient = Widget.extend(ServiceProviderMixin, {
         this.menu_dm = new concurrency.DropMisordered();
         this.action_mutex = new concurrency.Mutex();
         this.set('title_part', {"zopenerp": "Odoo"});
+        this.keyboardNavigation = new KeyboardNavigation();
```
[the remark this answers is not in the document]
Comment (VincentSchippefilt) [reply addons/web/static/src/js/chrome/abstract_web_client.js]: @alexkuhn is turning it into a mixin

File: addons/web/static/src/js/chrome/action_manager_act_window.js [hunk 43]
```diff
@@ -820,6 +821,20 @@ ActionManager.include({
         var addedFilters = action.searchView.updateFilters(data.newFilters, data.filtersToRemove);
         data.callback(addedFilters);
     },
+    /**
+     * @private
+     */
+    _onNavigationMove : function(event) {
```
[the remark this answers is not in the document]
Comment (VincentSchippefilt) [reply addons/web/static/src/js/chrome/action_manager_act_window.js]: done

File: addons/web/static/src/js/chrome/action_manager_act_window.js [hunk 44]
```diff
@@ -820,6 +821,20 @@ ActionManager.include({
...
+        switch(event.data.direction) {
+            case 'down' :
+                var currentController = this.getCurrentController().widget;
+                if (currentController.giveFocus) {
```
[the remark this answers is not in the document]
Comment (VincentSchippefilt) [reply addons/web/static/src/js/chrome/action_manager_act_window.js]: done

File: addons/web/static/src/js/chrome/abstract_web_client.js [hunk 45]
```diff
@@ -161,12 +164,17 @@ var AbstractWebClient = Widget.extend(ServiceProviderMixin, {
                 }
             }, 0);
         });
+        window.addEventListener('blur', function (e) { self.keyboardNavigation.hideAccessKeyOverlay(self.$el); });
+        this.$el.on('keydown', function (e) { self.keyboardNavigation.processKeyDown(e, self.$el); });
```
[the remark this answers is not in the document]
Comment (VincentSchippefilt) [reply addons/web/static/src/js/chrome/abstract_web_client.js]: @alexkuhn can you see if it makes sence in the mixin ?

File: addons/web/static/src/js/chrome/search_view.js [hunk 46]
```diff
@@ -417,10 +423,16 @@ var SearchView = Widget.extend({
         if (this.$buttons) {
             this.$buttons.toggle(!this.headless && is_visible && this.visible_filters);
         }
+        this.focus_input();
+    },
+    focus_input: function () {
```
[the remark this answers is not in the document]
Comment (VincentSchippefilt) [reply addons/web/static/src/js/chrome/search_view.js]: done

File: addons/web/static/src/js/chrome/search_view.js [hunk 47]
```diff
@@ -417,10 +423,16 @@ var SearchView = Widget.extend({
...
         if (!config.device.touch && config.device.size_class >= config.device.SIZES.SM) {
             this.$('input').focus();
         }
     },
+    on_attach_callback: function() {
```
[the remark this answers is not in the document]
Comment (VincentSchippefilt) [reply addons/web/static/src/js/chrome/search_view.js]: done

File: addons/web/static/src/js/core/dialog.js [hunk 48]
```diff
@@ -126,6 +128,25 @@ var Dialog = Widget.extend({
                     $.when(def).always(self.close.bind(self));
                 }
             });
+            $button.on('keydown', function(e) {
```
[the remark this answers is not in the document]
Comment (VincentSchippefilt) [reply addons/web/static/src/js/core/dialog.js]: done for keydown. Seems like the click needs a function to call as part of the click handler, so I will not refactor it.

File: addons/web/static/src/js/core/dialog.js [hunk 49]
```diff
@@ -148,7 +169,13 @@ var Dialog = Widget.extend({
         return (handler)? this._opened.then(handler) : this._opened;
     },
 
-    open: function () {
+    /**
+     * Show a dialog
+     * 
+     * @param {Object} options
+     * @param {Boolean} options.shouldFocusButtons  if true, put the focus on the first button primary when the dialog opens
```
[the remark this answers is not in the document]
Comment (VincentSchippefilt) [reply addons/web/static/src/js/core/dialog.js]: done

File: addons/web/static/src/js/core/dialog.js (hunk 30 above)
[the remark this answers is not in the document]
Comment (VincentSchippefilt) [reply addons/web/static/src/js/core/dialog.js]: done

File: addons/web/static/src/js/core/dialog.js [hunk 50]
```diff
@@ -184,11 +216,31 @@ var Dialog = Widget.extend({
...
+     * Manages the focus when the dialog closes. The default behavior is to set the focus on the top-most opened popup.
+     * The goal of this function is to be overridden by all children of the dialog class.
+     * 
+     * @returns: bool  should return true if the focus has already been set else false.
+     */
+    _focusOnClose: function() {
+        return false;
+    },
+
+    _onFocusControlButton:function(e) {
```
[the remark this answers is not in the document]
Comment (VincentSchippefilt) [reply addons/web/static/src/js/core/dialog.js]: done

File: addons/web/static/src/js/core/mixins.js [hunk 51]
```diff
@@ -331,6 +331,8 @@ var EventDispatcherMixin = _.extend({}, ParentedMixin, {
     },
     trigger_up: function (name, info) {
         var event = new OdooEvent(this, name, info);
+        // VSC: I am leaving this line commented on purpose. It is easier to debug events when they are logged.
```
[the remark this answers is not in the document]
Comment (VincentSchippefilt) [reply addons/web/static/src/js/core/mixins.js]: done

File: addons/web/static/src/js/fields/abstract_field.js [hunk 52]
```diff
@@ -299,6 +299,32 @@ var AbstractField = Widget.extend({
...
+     */
+    removeInvalidClass: function () {
+        this.$el.toggleClass('o_field_invalid',false);
+    },
+
+    /**
+     * Update the modifiers with the newest value. 
+     * Now this.attrs.modifiersValue can be used consistantly even with conditional modifiers
+     * 
+     * @param {Object} modifiers  the updated modifiers
```
[the remark this answers is not in the document]
Comment (VincentSchippefilt) [reply addons/web/static/src/js/fields/abstract_field.js]: done

File: addons/web/static/src/js/core/widget.js [hunk 53]
```diff
@@ -284,6 +284,15 @@ var Widget = core.Class.extend(mixins.PropertiesMixin, ServicesMixin, {
...
         return this;
     },
+    
+    /**
+     * Update the modifiers with the newest value. 
+     * Now this.attrs.modifiersValue can be used consistantly even with conditional modifiers
+     * 
+     * @param {Object} modifiers  the updated modifiers
+     */
+    updateModifiersValue: function(modifiers) {
```
[the remark this answers is not in the document]
Comment (VincentSchippefilt) [reply addons/web/static/src/js/core/widget.js]: it is actually required (crash on form_controller in the tests if I remove it)

File: addons/web/static/src/js/fields/abstract_field.js (hunk 4 above)
Comment (VincentSchippefilt) [reply addons/web/static/src/js/fields/abstract_field.js]: done

File: addons/web/static/src/js/fields/abstract_field.js (hunk 5 above)
Comment (VincentSchippefilt) [reply addons/web/static/src/js/fields/abstract_field.js]: done

File: addons/web/static/src/js/fields/abstract_field.js (hunk 6 above)
Comment (VincentSchippefilt) [reply addons/web/static/src/js/fields/abstract_field.js]: done

File: addons/web/static/src/js/fields/abstract_field.js (hunk 7 above)
Comment (VincentSchippefilt) [reply addons/web/static/src/js/fields/abstract_field.js]: done

File: addons/web/static/src/js/fields/relational_fields.js (hunk 9 above)
Comment (VincentSchippefilt) [reply addons/web/static/src/js/fields/relational_fields.js]: done

File: addons/web/static/src/js/views/form/form_controller.js (hunk 11 above)
Comment (VincentSchippefilt) [reply addons/web/static/src/js/views/form/form_controller.js]: done

File: addons/web/static/src/js/views/form/form_controller.js (hunk 12 above)
Comment (VincentSchippefilt) [reply addons/web/static/src/js/views/form/form_controller.js]: done

File: addons/web/static/src/js/views/form/form_controller.js (hunk 13 above)
Comment (VincentSchippefilt) [reply addons/web/static/src/js/views/form/form_controller.js]: done

File: addons/web/static/src/js/views/form/form_controller.js (hunk 14 above)
Comment (VincentSchippefilt) [reply addons/web/static/src/js/views/form/form_controller.js]: done

File: addons/web/static/src/js/views/form/form_renderer.js (hunk 15 above)
Comment (VincentSchippefilt) [reply addons/web/static/src/js/views/form/form_renderer.js]: done
